### PR TITLE
Add the design on long-term policy compliance history

### PR DESCRIPTION
**Pull request by mprahl**

### CODE REVIEW

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:124]: So now a service will be defined for the policy propagator.  In OCM it needs to be exposed over an ingress and in OCP it would have a route.  TLS termination might be different between the 2 scenarios.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:165]: Would it just always provide the `specHash`?  I expect the "spec" is only initialized on the hub by the propagator so managed clusters don't get to define a new spec.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:240]: So the entire spec is in all of these output entries?

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:285]: I guess I'm wondering if there's a difference between the hub managed cluster replicated spec and the spec that the managed cluster is sending here in the POST.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:262]: In the CSV I am guessing the spec would not be included and the message would be reformatted to not contain commas

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:165]: That's a good idea. The propagator *could* create entries for `spec` in the `policies` table so only the `specHash` needs to be provided from the managed cluster. That does simplify things. I'll update the doc with that.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:423]: My opinion is no.  That is not necessary to try to capture.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:44]: It's not clear that only one hub can write data to this database.  Could a customer with multiple hubs use the same database?  If so maybe the schema needs a hub reference too.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:124]: Yes, do you think it needs to be noted in the design?

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:240]: Yes, I think that's the easiest way to consume the data. Do you have other suggestions?

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:285]: This will be covered in the conversation in https://github.com/open-cluster-management-io/enhancements/pull/98#pullrequestreview-1548522146.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:240]: It feels a lot like the `managedFields` data for k8s resources.  It's nice to have it there and necessary to be able to get it, but most of the time you don't need to see it... Maybe an `includeSpec` option in the request would be nice to think about.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:124]: Yes, I think it would be valuable for clarity.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:187]: I'm just thinking about this permission. Should everyone that can `get` that ManagedCluster be able to see all the compliance events on all the policies on that cluster?

I don't see a great alternative (it seems like custom rbac verbs aren't a thing, otherwise `get-compliance-events` on the ManagedCluster would be ideal), but maybe a non-resource-url `/compliance-events` would help? Users would need a clusterrole for that (since non-resource-urls are cluster-scoped, I think), and additionally they would need `get` on the `ManagedClusters` they're asking for.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:122]: Customers may want to use ESO so having a secret with username and password directly in the secret instead of that URL format I think would be needed.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:321]: Why not always send the events to the embedded etcd first, and then always just POST them to the hub from that? I'm thinking that sending the event to the hub would be the slow part of the operation, so waiting for that to complete (or fail) would limit the number of policies the status-sync can handle at once.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:347]: Is a shared secret the best way to do this? I was expecting the same kind of authorization that is protecting the hub API endpoint (requiring "write access to the status subresource of the policies.policy.open-cluster-management.io resource").

The shared secret does make the embedded etcd simpler, since it doesn't need a proxy in front of it or anything...

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:423]: It depends on how hard it is to capture? I think a one-off looking through all the existing policies on the hub and posting compliance events for their histories could be ok. Just set the spec to unknown or something.

**Comment by mprahl** [comment]
I posted up some updates but haven't gotten to all the feedback yet. I'll work on this again tomorrow.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:44]: This schema allows the same database to be used for multiple Open Cluster Management hubs, though if managed cluster names overlap, it'd be less confusing if a separate database on the same Postgresql database be used.

I'll add this in the document.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md]: Yes, that was kept in by accident, though to address @gparvin's comment of storing the categories, controls, and standards, it is now required and the `policies` table was adjusted accordingly.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md]: I'll clarify this and the format can be seen in the example API calls b elow.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:354]: I was thinking to limit the chance that the message is lost since we aren't guaranteed a persistent volume but we can just prevent the poller/syncer goroutine from letting the pod shutdown until all compliance events are synced on the hub when embedded etcd is not backed by a persistent volume.

I change the design to have it like this.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:380]: Since the embedded etcd is an internal implementation detail, I'd like to avoid using the Kubernetes API for authorization since this adds an unnecessary additional load on the Kubernetes API and would slow recording the compliance events.

**Comment by mprahl** [comment]
@gparvin and @JustinKuli thanks for the feedback! This is ready for review again.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (yiraeChristineKim) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:235]: How about add `search` query? search with **words**

Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:235]: Good idea. I'll add the `message_includes` query parameter.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:300]: By default, it shows all the compliance events for all managed clusters the user has access to. You can filter with `?cluster=local-cluster&cluster=managed1`.

**Comment by mprahl** [comment]
@gparvin @JustinKuli @yiraeChristineKim, I just added diagrams for each "phase" to help clarify the proposed processes.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:301]: Thanks for pointing that out! It should be `clusters'` and not `cluster's`.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:474]: That's a good point on deleted clusters. I think you could still have a custom verb since `resourceNames` would either be not set or they would include the deleted clusters.

**Comment by mprahl** [comment]
@gparvin, @JustinKuli, and @yiraeChristineKim could you please review this again?

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (gparvin) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:476]: This can be further clarified in the threat model

**Comment by gparvin** [comment]
To make sure others can get comments considered adding a hold
/hold

**Review by gparvin — APPROVED** [review]
This looks good.  Regarding some of the comments around security (rbac/etcd usage/customer provided postgress and secret), I think one of the early work items needs to be creation of a threat model PR.  Can you make sure that is started early in dev?  Thanks!

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:34]: This seems good. I was more just wondering if there was a firm "we must support x" kind of requirement somewhere.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:78]: Do other SQL flavors allow filters like "if string x is found inside this column?" Like a substring match...

The way the policy already stores these is just as a comma-delimited list, since each annotation is a string, not a list. So if we could just put each annotation directly into these columns, that could be slightly neater on that side... and then regardless of the database flavor, it might still be possible to get the filtering.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:75]: What would happen if I update a policy, and add a new standard to it?

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (JustinKuli) [inline enhancements/sig-policy/98-long-term-compliance-history/README.md:304]: This report might be very large, right? The proposed GET `/api/v1/compliance-events` has pagination, but a report might have more than 100 entries (the maximum page size). For one "report" style API I worked with (as a consumer), I think I had to do a POST with the parameters for the report, the API responded with some kind of Accepted response, and gave a report ID that could be used later in a GET when the report is ready.

**Review by JustinKuli — COMMENTED** [review]
Overall looks good to me! Nothing we can't clarify while working on it.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:78]: @JustinKuli, I think a JSON column type may be better for portability to another database type. Postgres and MySQL both have it. SQLite has JSON functions that can parse text columns.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:75]: There would be a new `parent_policies` row. That's what has been the tricky part in this database schema design is we need to store what things were in that point in time and not what they currently are, so it does lead to some data duplication.

File: enhancements/sig-policy/98-long-term-compliance-history/README.md
Comment (mprahl) [reply enhancements/sig-policy/98-long-term-compliance-history/README.md:78]: After doing some more research, I don't think a JSON type is good for unique constraints. I like your idea of allowing substring checking for non-Postgresql databases.

**Review by JustinKuli — APPROVED** [review]

**Pull request merged**
